### ISSUE TITLE
Error: An item with the same key has already been added.

### ISSUE THREAD

**Issue by tpperlman** [2sic/2sxc#2127]
**I'm submitting a ...** 
[x] bug report => search github for a similar issue before submitting 

**...about**
[x] Content Types or data management

**Current behavior**
Following a 2SXC upgrade from 2SXC 10.27, when I add new data to a content type, my 2SXC app appears blank and the following error appears in the log:

```
AbsoluteURL:/Default.aspx
DefaultDataProvider:DotNetNuke.Data.SqlDataProvider, DotNetNuke
ExceptionGUID:2f1a01a4-1395-4136-bb70-cb1282dbbd28
AssemblyVersion:
PortalId:-1
UserId:-1
TabId:-1
RawUrl:
Referrer:
UserAgent:
ExceptionHash:0ZWSzFe4ulsLCicLeOaXYWQ8W98=
Message:An item with the same key has already been added.
StackTrace:
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at ToSic.Eav.Persistence.Efc.Efc11Loader.LoadEntities(AppState app, Int32[] entityIds) in C:\Projects\eav-server\ToSic.Eav.Persistence.EFC11\Efc11Loader_Entities.cs:line 71
   at ToSic.Eav.Persistence.Efc.Efc11Loader.<>c__DisplayClass8_0.<Update>b__0() in C:\Projects\eav-server\ToSic.Eav.Persistence.EFC11\Efc11Loader_App.cs:line 87
   at ToSic.Eav.Apps.AppState.Load(ILog parentLog, Action loader) in C:\Projects\eav-server\ToSic.Eav.Core\Apps\State\AppState_Load.cs:line 33
   at ToSic.Eav.Persistence.Efc.Efc11Loader.Update(AppState app, AppStateLoadSequence startAt, Int32[] entityIds, ILog parentLog) in C:\Projects\eav-server\ToSic.Eav.Persistence.EFC11\Efc11Loader_App.cs:line 94
   at ToSic.Eav.Persistence.Efc.Efc11Loader.AppState(Int32 appId, Int32[] entityIds, ILog parentLog) in C:\Projects\eav-server\ToSic.Eav.Persistence.EFC11\Efc11Loader_App.cs:line 47
   at ToSic.Eav.Caching.AppsCacheBase.GetOrBuild(IAppIdentity appIdentity, String primaryLanguage) in C:\Projects\eav-server\ToSic.Eav.Core\Caching\Apps\AppsCacheBase.cs:line 111
   at ToSic.Eav.Apps.App..ctor(Int32 zoneId, Int32 appId, Boolean allowSideEffects, Func`2 buildConfiguration, ILog parentLog, String logMsg) in C:\Projects\eav-server\ToSic.Eav.AppEngine\App.cs:line 49
   at ToSic.Eav.Apps.App..ctor(IAppEnvironment env, ITenant tenant, Int32 zoneId, Int32 appId, Boolean allowSideEffects, Func`2 buildConfiguration, ILog parentLog) in C:\Projects\eav-server\ToSic.Eav.AppEngine\App_Environment.cs:line 33
   at ToSic.Sxc.Blocks.BlockFromModule..ctor(IContainer container, ILog parentLog, ITenant tenant, IEnumerable`1 overrideParams) in C:\Projects\2sxc-dnn742\Website\DesktopModules\ToSIC_SexyContent\ToSic.Sxc\Blocks\BlockFromModule.cs:line 70
   at ToSic.SexyContent.View.get_BlockBuilder() in C:\Projects\2sxc-dnn742\Website\DesktopModules\ToSIC_SexyContent\View.ascx.cs:line 24
   at ToSic.SexyContent.View.InitModuleActions() in C:\Projects\2sxc-dnn742\Website\DesktopModules\ToSIC_SexyContent\View.ascx.menu.cs:line 42
   at ToSic.SexyContent.View.get_ModuleActions() in C:\Projects\2sxc-dnn742\Website\DesktopModules\ToSIC_SexyContent\View.ascx.menu.cs:line 29
InnerMessage:
InnerStackTrace:
Source:mscorlib
FileName:
FileLineNumber:0
FileColumnNumber:0
Method:
```

**Minimal reproduction of the problem with instructions**
This is an existing app developed a couple years back, and it doesn't appear to happen in any of the other 2sxc apps in my site.

* **2sxc version(s):** 11.03.00
* **Browser:** all
* **DNN:** 09.06.01
* **Language:** any/all

**Anything you would like to add**
I have tried to scour the 2SXC tables and can't see any duplicate entries. The error also blocks me from opening the 2sxc admin panel for any other apps with the message:

```
Had an error talking to the server (status 400).
Message: Bad Request
Detail: An item with the same key has already been added.
```

When I add the new data, it appears fine. But when the app pool is reset and/or the cache cleared then the above error appears. Is there any way to trace the source of the duplicate entry?

**Comment by tpperlman** [comment]
Additionally, I looked at the ToSIC_EAV_Apps table and don't see any duplicates, nor any duplicate 2SexyContent-App entries in the ToSIC_EAV_Entities table.

**Comment by 10PoundGorilla-ChrisBehling** [comment]
We are also experiencing this issue on a DNN platform upgrade while also upgrading 2sxc

**Original Specs:**
DNN: 9.02.02
2sxc: 9.43.01

**New Specs:**
DNN: 9.03.02
2sxc: 11.03.0

The "New Specs" are a middle-version upgrade before we jump to DNN 9.06.02

**Comment by tpperlman** [comment]
@10PoundGorilla-ChrisBehling having not gotten a reply, I had to export all my data, delete/recreate my content types and update queries to get it back. Still struggling through some issues... I couldn't fine any duplicate GUIDs, attributes or attribute sets in the database that would cause the issue. It seems to have been related to linking entities somehow.

**Comment by PagnoDunadan** [comment]
It's time of summer vacations and unfortunately I don't know how backend works to help you out. Daniel won't be back for at least another week.

**Comment by tpperlman** [comment]
@PagnoDunadan thank you... everyone deserves a break, especially these days. I'll try and work my way through it, but with the quantity of records importing takes forever.

**Comment by tpperlman** [comment]
The biggest problem with this error is that our upgrades appeared well. Then when the app pool recycled or the cache was cleared it would appear. We didn't realize until many days later when we added/updated data of this.

If anyone else does upgrade from an earlier release to version 10.2x+or so be sure to clear cache and recycle the app after adding/editing data to ensure that you aren't affected.

**Comment by nicpitsch** [comment]
> Additionally, I looked at the ToSIC_EAV_Apps table and don't see any duplicates, nor any duplicate 2SexyContent-App entries in the ToSIC_EAV_Entities table.

Did you check the tables `ToSIC_EAV_Attribute` and `ToSIC_EAV_AttributeSets` according to Daniel's comment here: https://github.com/2sic/2sxc/issues/1600#issuecomment-408890398?

**Comment by tpperlman** [comment]
Thansk @nicpitsch, yes I did. I went through all the tables painstakingly over 5 days to try and find any duplicates or errors and couldn't.

I had been hoping that 2SIC had some scripts to identify and resolve corruption issues, but I ended up creating the stored procedure below to help cleanup the tables. Unfortunately the error kept creeping back in so I had to recreate the whole app and the pain of slow data imports to overcome the issue.

```
-- UNTESTED CODE
-- USE THIS CODE AT YOUR OWN RISK

DECLARE @AppId int = XX
DECLARE @NumDuplicates int

SELECT @NumDuplicates = COUNT(AttributeSetId)
FROM ToSIC_EAV_AttributeSets
WHERE Name in 
(
	SELECT [Name]
	FROM ToSIC_EAV_AttributeSets
	where appid = @AppId
	GROUP BY [Name]
	HAVING COUNT(*) > 1
)

IF @NumDuplicates < 1 BEGIN

	PRINT 'There is no data to clean up.'

END ELSE BEGIN

	-- UPDATE ENTITIES

	UPDATE ToSIC_EAV_Entities
		SET AttributeSetID = NewAttributeSetID
	FROM ToSIC_EAV_Entities
	INNER JOIN (
		SELECT ToSIC_EAV_AttributeSets.Name, ToSIC_EAV_Attributes.StaticName, MIN(ToSIC_EAV_AttributeSets.AttributeSetID) as OldAttributeSetID, MAX(ToSIC_EAV_AttributeSets.AttributeSetID) as NewAttributeSetID
		FROM ToSIC_EAV_AttributeSets
		INNER JOIN ToSIC_EAV_AttributesInSets ON ToSIC_EAV_AttributesInSets.AttributeSetID = ToSIC_EAV_AttributeSets.AttributeSetID
		INNER JOIN ToSIC_EAV_Attributes ON ToSIC_EAV_AttributesInSets.AttributeID = ToSIC_EAV_Attributes.AttributeID
		INNER JOIN (
			SELECT AttributeSetId
			FROM ToSIC_EAV_AttributeSets
			WHERE Name in 
			(
				SELECT [Name]
				FROM ToSIC_EAV_AttributeSets
				where appid = @AppId
				GROUP BY [Name]
				HAVING COUNT(*) > 1
			)
		) as dupes ON dupes.AttributeSetID = ToSIC_EAV_AttributeSets.AttributeSetID
		GROUP BY ToSIC_EAV_AttributeSets.Name, ToSIC_EAV_Attributes.StaticName
	) as d ON ToSIC_EAV_Entities.AttributeSetID = d.OldAttributeSetID
	WHERE ToSIC_EAV_Entities.AttributeSetID = OldAttributeSetID


	-- UPDATE ENTITY RELATIONSHIPS

	UPDATE ToSIC_EAV_EntityRelationships
		SET AttributeID = NewAttributeID
	FROM ToSIC_EAV_EntityRelationships
	INNER JOIN (
		SELECT ToSIC_EAV_AttributeSets.Name, ToSIC_EAV_Attributes.StaticName, MIN(ToSIC_EAV_Attributes.AttributeID) as OldAttributeID, MAX(ToSIC_EAV_Attributes.AttributeID) as NewAttributeID
		FROM ToSIC_EAV_AttributeSets
		INNER JOIN ToSIC_EAV_AttributesInSets ON ToSIC_EAV_AttributesInSets.AttributeSetID = ToSIC_EAV_AttributeSets.AttributeSetID
		INNER JOIN ToSIC_EAV_Attributes ON ToSIC_EAV_AttributesInSets.AttributeID = ToSIC_EAV_Attributes.AttributeID
		INNER JOIN ToSIC_EAV_EntityRelationships ON ToSIC_EAV_EntityRelationships.AttributeID = ToSIC_EAV_Attributes.AttributeID
		INNER JOIN (
			SELECT AttributeSetId
			FROM ToSIC_EAV_AttributeSets
			WHERE Name in 
			(
				SELECT [Name]
				FROM ToSIC_EAV_AttributeSets
				where appid = @AppId
				GROUP BY [Name]
				HAVING COUNT(*) > 1
			)
		) as dupes ON dupes.AttributeSetID = ToSIC_EAV_AttributeSets.AttributeSetID
		GROUP BY ToSIC_EAV_AttributeSets.Name, ToSIC_EAV_Attributes.StaticName
	) as d ON ToSIC_EAV_EntityRelationships.AttributeID = d.OldAttributeID
	WHERE ToSIC_EAV_EntityRelationships.AttributeID = OldAttributeID


	-- UPDATE VALUES

	UPDATE ToSIC_EAV_Values
		SET AttributeID = NewAttributeID
	FROM ToSIC_EAV_Values
	INNER JOIN (
		SELECT ToSIC_EAV_AttributeSets.Name, ToSIC_EAV_Attributes.StaticName, MIN(ToSIC_EAV_Attributes.AttributeID) as OldAttributeID, MAX(ToSIC_EAV_Attributes.AttributeID) as NewAttributeID
		FROM ToSIC_EAV_AttributeSets
		INNER JOIN ToSIC_EAV_AttributesInSets ON ToSIC_EAV_AttributesInSets.AttributeSetID = ToSIC_EAV_AttributeSets.AttributeSetID
		INNER JOIN ToSIC_EAV_Attributes ON ToSIC_EAV_AttributesInSets.AttributeID = ToSIC_EAV_Attributes.AttributeID
		INNER JOIN ToSIC_EAV_Values ON ToSIC_EAV_Values.AttributeID = ToSIC_EAV_Attributes.AttributeID
		INNER JOIN (
			SELECT AttributeSetId
			FROM ToSIC_EAV_AttributeSets
			WHERE Name in 
			(
				SELECT [Name]
				FROM ToSIC_EAV_AttributeSets
				where appid = @AppId
				GROUP BY [Name]
				HAVING COUNT(*) > 1
			)
		) as dupes ON dupes.AttributeSetID = ToSIC_EAV_AttributeSets.AttributeSetID
		GROUP BY ToSIC_EAV_AttributeSets.Name, ToSIC_EAV_Attributes.StaticName
	) as d ON ToSIC_EAV_Values.AttributeID = d.OldAttributeID
	WHERE ToSIC_EAV_Values.AttributeID = OldAttributeID


	-- GET ATTRIBUTES

	SELECT ToSIC_EAV_AttributeSets.Name, ToSIC_EAV_Attributes.StaticName, MIN(ToSIC_EAV_AttributesInSets.AttributeID) as OldAttributeID, MAX(ToSIC_EAV_AttributesInSets.AttributeID) as NewAttributeID
	INTO #Attributes
	FROM ToSIC_EAV_AttributeSets
	INNER JOIN ToSIC_EAV_AttributesInSets ON ToSIC_EAV_AttributesInSets.AttributeSetID = ToSIC_EAV_AttributeSets.AttributeSetID
	INNER JOIN ToSIC_EAV_Attributes ON ToSIC_EAV_AttributesInSets.AttributeID = ToSIC_EAV_Attributes.AttributeID
	INNER JOIN (
		SELECT AttributeSetID
		FROM ToSIC_EAV_AttributeSets
		WHERE Name in 
		(
			SELECT [Name]
			FROM ToSIC_EAV_AttributeSets
			where appid = @AppId
			GROUP BY [Name]
			HAVING COUNT(*) > 1
		)
	) as dupes ON dupes.AttributeSetID = ToSIC_EAV_AttributeSets.AttributeSetID
	GROUP BY ToSIC_EAV_AttributeSets.Name, ToSIC_EAV_Attributes.StaticName


	-- GET ATTRIBUTE SETS

	SELECT MIN(AttributeSetId) as OldAttributeSetId
	INTO #AttributeSets
	FROM ToSIC_EAV_AttributeSets
	WHERE Name in 
	(
		SELECT [Name]
		FROM ToSIC_EAV_AttributeSets
		where appid = @AppId
		GROUP BY [Name]
		HAVING COUNT(*) > 1
	)
	GROUP BY ToSIC_EAV_AttributeSets.Name


	-- DROP CONSTRAINTS

	IF  EXISTS (SELECT * FROM sys.objects WHERE object_id = OBJECT_ID(N'[dbo].[ToSIC_EAV_EntityRelationships]') AND type in (N'U'))
		ALTER TABLE [dbo].[ToSIC_EAV_EntityRelationships] DROP CONSTRAINT IF EXISTS [FK_ToSIC_EAV_EntityRelationships_ToSIC_EAV_Attributes]

	IF  EXISTS (SELECT * FROM sys.objects WHERE object_id = OBJECT_ID(N'[dbo].[ToSIC_EAV_Entities]') AND type in (N'U'))
		ALTER TABLE [dbo].[ToSIC_EAV_Entities] DROP CONSTRAINT IF EXISTS [FK_ToSIC_EAV_Entities_ToSIC_EAV_AttributeSets]


	---- DELETE ATTRIBUTES IN SETS

	DELETE FROM ToSIC_EAV_AttributesInSets WHERE AttributeSetID in (SELECT OldAttributeSetId FROM #AttributeSets)


	---- DELETE ATTTRIBUTES

	DELETE FROM ToSIC_EAV_Attributes WHERE AttributeID in (SELECT OldAttributeID FROM #Attributes)


	---- DELETE ATTRIBUTE SETS

	DELETE FROM ToSIC_EAV_AttributeSets WHERE AttributeSetID in (SELECT OldAttributeSetId FROM #AttributeSets)


	-- CLEANUP

	DROP TABLE #Attributes

	DROP TABLE #AttributeSets


	-- RECREATE CONSTRAINTS

	ALTER TABLE [dbo].[ToSIC_EAV_EntityRelationships]  WITH NOCHECK ADD  CONSTRAINT [FK_ToSIC_EAV_EntityRelationships_ToSIC_EAV_Attributes] FOREIGN KEY([AttributeID])
	REFERENCES [dbo].[ToSIC_EAV_Attributes] ([AttributeID])

	ALTER TABLE [dbo].[ToSIC_EAV_EntityRelationships] CHECK CONSTRAINT [FK_ToSIC_EAV_EntityRelationships_ToSIC_EAV_Attributes]


	ALTER TABLE [dbo].[ToSIC_EAV_Entities]  WITH NOCHECK ADD  CONSTRAINT [FK_ToSIC_EAV_Entities_ToSIC_EAV_AttributeSets] FOREIGN KEY([AttributeSetID])
	REFERENCES [dbo].[ToSIC_EAV_AttributeSets] ([AttributeSetID])

	ALTER TABLE [dbo].[ToSIC_EAV_Entities] CHECK CONSTRAINT [FK_ToSIC_EAV_Entities_ToSIC_EAV_AttributeSets]

END
```

**Comment by tpperlman** [comment]
Closing the issue since I have resolved my problem by recreating the app.

**Comment by tpperlman** [comment]
This issue has mysteriously returned and I can't diagnose it. Adding new records through the UI works, but hours later (app/cache refresh?) the "An item with the same key has already been added" message appears in the log and the modules are all empty.

I spent 30+ hours diagnosing/recreating the app when it happened last time, and I hope that someone can help me diagnose it this time.

**Comment by tpperlman** [comment]
@2sicProgrammers @iJungleboy do you guys have any advice? I am at a total loss of what to do.

**Comment by iJungleboy** [comment]
@tpperlman Would love to help, sorry to see you suffer 😢

It's been a while so I hope I'm not writing things you already tried. 

I would really check the insights, where the app is being built from the DB 
![image](https://user-images.githubusercontent.com/5321186/92236789-6ee23b00-eeb6-11ea-9e20-57dffc6f3a35.png)

![image](https://user-images.githubusercontent.com/5321186/92236823-83263800-eeb6-11ea-8f23-b25df4f55169.png)

![image](https://user-images.githubusercontent.com/5321186/92236847-8c170980-eeb6-11ea-9e43-1fe4941cbd3e.png)

you should see a lot of 
![image](https://user-images.githubusercontent.com/5321186/92236877-9d601600-eeb6-11ea-8905-143168253495.png)

but there where it ends, it ran into something.

**Comment by tpperlman** [comment]
Nothing stands out to me... these are the app-state logs when I load a page with the affected modules that appear blank:

```
Log for app-state[2]

    App.St-8[d5] - start build AppState 8 C#
    App.St-8[d5] - cache reset #1 to stamp 637348066302701680 C#
    App.St-8[d5] - Load() zone/app:2/8 C#
        App.St-8[d5] - app loading start C#
        App.St-8[d5]DB.EFLoad[45] - get app data package for a#8, startAt: Start, ids only:False C#
        App.St-8[d5]DB.EFLoad[45] - Update() C#
            App.St-8[d5]App.MDMan[59] - initialize C#
            DB.EFLoad[45] - PreloadAppPath() = Listings C#
                DB.EFLoad[45] - GetRawEntities(app: 8, ids: 0, filter: False) = ok C#
                DB.EFLoad[45] - app Entity found - this json: {"_":{"V":1},"Entity":{"Id":16809,"Version":1,"Guid":"5f674c9b-c384-4751-9e65-027edb0c7e9f","Type":{"Name":"2SexyContent-App","Id":"2SexyContent-App"},"Attributes":{"String":{"DisplayName":{"*":"Listings"},"Folder":{"*":"Listings"},"OriginalId":{"*":""},"Version":{"*":"00.00.01"}}},"For":{"Target":"App","Number":8}}} C#
                Jsn.Serlzr[4e] - UnpackAndTestGenericJsonV1() = ok C#
                Jsn.Serlzr[4e] - Deserialize(guid: 5f674c9b-c384-4751-9e65-027edb0c7e9f; allowDynamic:True skipUnknown:True) = ok C#
                    Jsn.Serlzr[4e] - this is metadata; will construct 'For' object. Type: App C#
                    Jsn.Serlzr[4e] - build entity C#
                    Jsn.Serlzr[4e] - BuildAttribsOfKnownType() = ok C#
            DB.EFLoad[45] - LoadContentTypesIntoLocalCache() = ⌚ 26.2741ms C#
            DB.EFLoad[45] - LoadExtensionsTypesAndMerge() = no app file types ⌚ 0.217ms C#
                DB.EFLoad[45] - InitFileSystemContentTypes() = ok C#
                    Dnn.AppStf[ce] - .ctor(8, Listings, ...) C#
                        Dnn.AppStf[ce] - System path:T:\Client Temp\FISAR\WWW\Portals\0\2sxc\Listings\system C#
                    Dnn.AppStf[ce] - ContentTypes() = ok C#
                        Dnn.AppStf[ce] - Found 0 extensions with .data folder C#
            App.St-8[d5] - InitContentTypes() init content types 14= ok ⌚ 0.0759ms C#
                App.St-8[d5] - BuildCacheForTypesByName() build cache for type names for 14 items= ok ⌚ 0.0353ms C#
            DB.EFLoad[45] - timers types:00:00:00.0266056 C#
            DB.EFLoad[45] - LoadEntities(8, 0) C#
                App.St-8[d5] - remove all items C#
                DB.EFLoad[45] - GetRawEntities(app: 8, ids: 0, filter: False) = ok C#
                DB.EFLoad[45] - Found 10037 raw entities in 53ms - chunked into 3 chunks C#
                DB.EFLoad[45] - GetRelatedEntities(app: 8, ids: 5000) = ok C#
                DB.EFLoad[45] - GetRelatedEntities(app: 8, ids: 5000) = ok C#
                DB.EFLoad[45] - GetRelatedEntities(app: 8, ids: 37) = ok C#
            end of log 
```

```
Log for app-state[3]

    App.St-8[64] - start build AppState 8 C#
    App.St-8[64] - cache reset #1 to stamp 637348066304541767 C#
    App.St-8[64] - Load() zone/app:2/8 C#
        App.St-8[64] - app loading start C#
        App.St-8[64]DB.EFLoad[71] - get app data package for a#8, startAt: Start, ids only:False C#
        App.St-8[64]DB.EFLoad[71] - Update() C#
            App.St-8[64]App.MDMan[79] - initialize C#
            DB.EFLoad[71] - PreloadAppPath() = Listings C#
                DB.EFLoad[71] - GetRawEntities(app: 8, ids: 0, filter: False) = ok C#
                DB.EFLoad[71] - app Entity found - this json: {"_":{"V":1},"Entity":{"Id":16809,"Version":1,"Guid":"5f674c9b-c384-4751-9e65-027edb0c7e9f","Type":{"Name":"2SexyContent-App","Id":"2SexyContent-App"},"Attributes":{"String":{"DisplayName":{"*":"Listings"},"Folder":{"*":"Listings"},"OriginalId":{"*":""},"Version":{"*":"00.00.01"}}},"For":{"Target":"App","Number":8}}} C#
                Jsn.Serlzr[74] - UnpackAndTestGenericJsonV1() = ok C#
                Jsn.Serlzr[74] - Deserialize(guid: 5f674c9b-c384-4751-9e65-027edb0c7e9f; allowDynamic:True skipUnknown:True) = ok C#
                    Jsn.Serlzr[74] - this is metadata; will construct 'For' object. Type: App C#
                    Jsn.Serlzr[74] - build entity C#
                    Jsn.Serlzr[74] - BuildAttribsOfKnownType() = ok C#
            DB.EFLoad[71] - LoadContentTypesIntoLocalCache() = ⌚ 25.9431ms C#
            DB.EFLoad[71] - LoadExtensionsTypesAndMerge() = no app file types ⌚ 0.1335ms C#
                DB.EFLoad[71] - InitFileSystemContentTypes() = ok C#
                    Dnn.AppStf[48] - .ctor(8, Listings, ...) C#
                        Dnn.AppStf[48] - System path:T:\Client Temp\FISAR\WWW\Portals\0\2sxc\Listings\system C#
                    Dnn.AppStf[48] - ContentTypes() = ok C#
                        Dnn.AppStf[48] - Found 0 extensions with .data folder C#
            App.St-8[64] - InitContentTypes() init content types 14= ok ⌚ 0.0658ms C#
                App.St-8[64] - BuildCacheForTypesByName() build cache for type names for 14 items= ok ⌚ 0.0345ms C#
            DB.EFLoad[71] - timers types:00:00:00.0261692 C#
            DB.EFLoad[71] - LoadEntities(8, 0) C#
                App.St-8[64] - remove all items C#
                DB.EFLoad[71] - GetRawEntities(app: 8, ids: 0, filter: False) = ok C#
                DB.EFLoad[71] - Found 10037 raw entities in 55ms - chunked into 3 chunks C#
                DB.EFLoad[71] - GetRelatedEntities(app: 8, ids: 5000) = ok C#
                DB.EFLoad[71] - GetRelatedEntities(app: 8, ids: 5000) = ok C#
                DB.EFLoad[71] - GetRelatedEntities(app: 8, ids: 37) = ok C#
            end of log 

```
```
Log for app-state[4]

    App.St-8[f6] - start build AppState 8 C#
    App.St-8[f6] - cache reset #1 to stamp 637348066306330300 C#
    App.St-8[f6] - Load() zone/app:2/8 C#
        App.St-8[f6] - app loading start C#
        App.St-8[f6]DB.EFLoad[6f] - get app data package for a#8, startAt: Start, ids only:False C#
        App.St-8[f6]DB.EFLoad[6f] - Update() C#
            App.St-8[f6]App.MDMan[dd] - initialize C#
            DB.EFLoad[6f] - PreloadAppPath() = Listings C#
                DB.EFLoad[6f] - GetRawEntities(app: 8, ids: 0, filter: False) = ok C#
                DB.EFLoad[6f] - app Entity found - this json: {"_":{"V":1},"Entity":{"Id":16809,"Version":1,"Guid":"5f674c9b-c384-4751-9e65-027edb0c7e9f","Type":{"Name":"2SexyContent-App","Id":"2SexyContent-App"},"Attributes":{"String":{"DisplayName":{"*":"Listings"},"Folder":{"*":"Listings"},"OriginalId":{"*":""},"Version":{"*":"00.00.01"}}},"For":{"Target":"App","Number":8}}} C#
                Jsn.Serlzr[27] - UnpackAndTestGenericJsonV1() = ok C#
                Jsn.Serlzr[27] - Deserialize(guid: 5f674c9b-c384-4751-9e65-027edb0c7e9f; allowDynamic:True skipUnknown:True) = ok C#
                    Jsn.Serlzr[27] - this is metadata; will construct 'For' object. Type: App C#
                    Jsn.Serlzr[27] - build entity C#
                    Jsn.Serlzr[27] - BuildAttribsOfKnownType() = ok C#
            DB.EFLoad[6f] - LoadContentTypesIntoLocalCache() = ⌚ 26.7404ms C#
            DB.EFLoad[6f] - LoadExtensionsTypesAndMerge() = no app file types ⌚ 0.1517ms C#
                DB.EFLoad[6f] - InitFileSystemContentTypes() = ok C#
                    Dnn.AppStf[44] - .ctor(8, Listings, ...) C#
                        Dnn.AppStf[44] - System path:T:\Client Temp\FISAR\WWW\Portals\0\2sxc\Listings\system C#
                    Dnn.AppStf[44] - ContentTypes() = ok C#
                        Dnn.AppStf[44] - Found 0 extensions with .data folder C#
            App.St-8[f6] - InitContentTypes() init content types 14= ok ⌚ 0.0692ms C#
                App.St-8[f6] - BuildCacheForTypesByName() build cache for type names for 14 items= ok ⌚ 0.0355ms C#
            DB.EFLoad[6f] - timers types:00:00:00.0269975 C#
            DB.EFLoad[6f] - LoadEntities(8, 0) C#
                App.St-8[f6] - remove all items C#
                DB.EFLoad[6f] - GetRawEntities(app: 8, ids: 0, filter: False) = ok C#
                DB.EFLoad[6f] - Found 10037 raw entities in 56ms - chunked into 3 chunks C#
                DB.EFLoad[6f] - GetRelatedEntities(app: 8, ids: 5000) = ok C#
                DB.EFLoad[6f] - GetRelatedEntities(app: 8, ids: 5000) = ok C#
                DB.EFLoad[6f] - GetRelatedEntities(app: 8, ids: 37) = ok C#
            end of log 
```

**Comment by tpperlman** [comment]
I've been doing a database compare (before/after adding new data item) to see where the issue may be from, but I am not seeing where the duplicate entry is coming from:

![2sxc](https://user-images.githubusercontent.com/1871915/92243814-10a26180-ee90-11ea-83d1-daed9f853a3c.jpg)

**Comment by iJungleboy** [comment]
This is hard and time consuming to do in chat... if it's for a company or something, can you get a small 2h Senior support budget ? I think we should be able to solve it more quickly that way. https://2sxc.org/contact

**Comment by tpperlman** [comment]
I will ask, business is a small mom-and-pop so that might be a hard sell. What if it ends up being an upgrade bug in 2sxc? I wasn't alone in upgrading from 2SXC ~v10 to ~v11 and this case has me afraid to do any further upgrades across sites running ~v10.

**Comment by iJungleboy** [comment]
@tpperlman as of now we don't know of anybody else having this issue, so I cannot say what causes it or if it's a bug in 2sxc or not. But since we don't make money selling 2sxc we cannot do personal support or prioritize any requests for free. As it's open source, people are welcome to contribute fixes etc. - but if you need our help asap we must charge for it.

**Comment by tpperlman** [comment]
@10PoundGorilla-ChrisBehling (above) was also having the issue following an upgrade. I completely understand your position and appreciate your contribution to the community. I am awaiting feedback from the customer... but nonetheless I am afraid to upgrade any of our many other sites running 2SXC because of it.

**Comment by iJungleboy** [comment]
BTW: I just checked the code and it seems to occur when loading relationships. Not sure if that helps.

**Comment by iJungleboy** [comment]
I just created a custom build for you which should provide more error details. As I can't really test it you'll have to check if it helps. Pls install this on a copy or similar of the site and check the insights logs. 
[2sic_2SexyContent_11.05.01_Install.zip](https://github.com/2sic/2sxc/files/5175655/2sic_2SexyContent_11.05.01_Install.zip)

**Comment by tpperlman** [comment]
Thank you for that, I am in the process of trying an export/import app to see if the problem carries through that. There are about 10,000 entities, so that will take some time. Afterwards, I'll restore from backup and try with the new build. I too felt it had to do with relationships and had rebuilt some of the data types as a result. But the problem appears out of nowhere, and the data types for the relationship haven't changed. All we've been doing is adding data items.

**Comment by iJungleboy** [comment]
BTW: I think I may have figured it out. 

Basically the SQL statement loads records in batches of 5'000 - that's to ensure we don't run into timeouts. This allows 2sxc apps to have more than 100'000 items and still work great. 

Since relationships have 2 parts, it loads "relationship-records" where an id is either parent or child. I believe what's happening is that it find the identical relationship in 2 different batches (chunks), pre-processes it, and when it should merge the chunks, it runs into issues. 

I should have another build for you fixing this in a few minutes.

**Comment by iJungleboy** [comment]
Try this on a copied site - I hope it will solve all your problems...
[2sic_2SexyContent_11.05.01_Install.zip](https://github.com/2sic/2sxc/files/5175817/2sic_2SexyContent_11.05.01_Install.zip)

**Comment by tpperlman** [comment]
That was the magic!! Without being able to look into the internals I never would have seen that. Your explanation makes total sense. Thank you SO much for the insight. Is this considered a dev build I should apply on the production server or should I await a new build? Have a great weekend, and have a drink on me!!

**Comment by iJungleboy** [comment]
@tpperlman your welcome - but please do consider motivating your customer to pay for 2 hours - it did take me about 1:30 so...

**Comment by iJungleboy** [comment]
in general I think you can use this release - it's almost the same as 11.05 - but that one just got out today so use some caution.

**Comment by iJungleboy** [comment]
Just fyi: we discovered a follow-up bug because of this #2201 which we just fixed in 11.07.03 LTS

**Comment by tpperlman** [comment]
@iJungleboy thank you very much for the follow up. We haven't experienced any further issues, but I will upgrade.
